packages/@lwc/engine/src/framework/wiring.ts

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { assert, isUndefined, ArrayPush, getOwnPropertyNames } from '@lwc/shared';
+import { ComponentInterface } from './component';
+import { componentValueMutated, ReactiveObserver } from './mutation-tracker';
+import { VM, runWithBoundaryProtection } from './vm';
+import { invokeComponentCallback } from './invoker';
+import { dispatchEvent } from '../env/dom';
+
+const WireMetaMap: Map<PropertyDescriptor, WireDef> = new Map();
+function noop(): void {}
+
+function createFieldDataCallback(vm: VM, name: string) {
+    const { cmpFields } = vm;
+    return (value: any) => {
+        if (value !== vm.cmpFields[name]) {
+            // storing the value in the underlying storage
+            cmpFields[name] = value;
+
+            componentValueMutated(vm, name);
+        }
+    };
+}
+
+function createMethodDataCallback(vm: VM, method: (data: any) => any) {
+    return (value: any) => {
+        // dispatching new value into the wired method
+        invokeComponentCallback(vm, method, [value]);
+    };
+}
+
+function createConfigWatcher(
+    vm: VM,
+    wireDef: WireDef,
+    callbackWhenConfigIsReady: (newConfig: ConfigValue) => void
+) {
+    const { component } = vm;
+    const { configCallback } = wireDef;
+    let hasPendingConfig: boolean = false;
+    // creating the reactive observer for reactive params when needed
+    const ro = new ReactiveObserver(() => {
+        if (hasPendingConfig === false) {
+            hasPendingConfig = true;
+            // collect new config in the micro-task
+            Promise.resolve().then(() => {
+                hasPendingConfig = false;
+                // resetting current reactive params
+                ro.reset();
+                // dispatching a new config due to a change in the configuration
+                callback();
+            });
+        }
+    });
+    const callback = () => {
+        let config: ConfigValue;
+        ro.observe(() => (config = configCallback(component)));
+        // eslint-disable-next-line lwc-internal/no-invalid-todo
+        // TODO: dev-mode validation of config based on the adapter.configSchema
```

Comment:
@caridy what do you mean by this todo? we have a new lint rule in place that states that for each todo in the code, there should be an open issue in the repo.

that being said, i don't know how we can proceed, because it does not become a real issue until the pr is merged. @pmdartus will have thoughts on this :), maybe make it a requirement for a pr to be merged?

Reply:
what I mean here is that `adapter.configSchema` describes what the adapter expects to be provided, while the compiler wire decorator configuration function produces a bag of configuration values, they must match, at at this point is where we have both at hand for the runtime validation. I suspect for example that we can provide nice errors like: 

* the adapter is expecting `foo` configuration, but that is missing in the wire declarative config for field `x`.

that's just an example, but the point here is that you cannot do during build time because you don't have the adapter, but at runtime you have both.

I'm fine with this being done in another PR, in fact I prefer that to be the case. Just create the issue, and update this todo.